File: backend/internal/core/github/oauth.go
// Copyright (C) 2025 Austin Beattie
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

package github

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// OAuthClientID is the OAuth client ID for Octobud - this is public and safe to embed
const OAuthClientID = "Ov23ct48TupzSLIkHEzO"

// Error definitions.
var (
	ErrAuthorizationPending = errors.New("authorization_pending")
	ErrSlowDown             = errors.New("slow_down")
	ErrExpiredToken         = errors.New("expired_token")
	ErrAccessDenied         = errors.New("access_denied")
)

// DeviceCodeResponse represents the response from GitHub's device code endpoint
type DeviceCodeResponse struct {
	DeviceCode      string `json:"device_code"`
	UserCode        string `json:"user_code"`
	VerificationURI string `json:"verification_uri"`
	ExpiresIn       int    `json:"expires_in"`
	Interval        int    `json:"interval"`
}

// AccessTokenResponse represents the response from GitHub's access token endpoint
type AccessTokenResponse struct {
	AccessToken string `json:"access_token"`
	TokenType   string `json:"token_type"`
	Scope       string `json:"scope"`
	Error       string `json:"error,omitempty"`
}

// OAuthClient handles GitHub OAuth Device Flow operations
type OAuthClient struct {
	httpClient *http.Client
	clientID   string
}

// NewOAuthClient creates a new OAuth client
func NewOAuthClient() *OAuthClient {
	return &OAuthClient{
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
		clientID: OAuthClientID,
	}
}

// RequestDeviceCode initiates the Device Flow by requesting a device code
func (c *OAuthClient) RequestDeviceCode(ctx context.Context) (*DeviceCodeResponse, error) {
	data := url.Values{}
	data.Set("client_id", c.clientID)
	data.Set("scope", "notifications repo")

	req, err := http.NewRequestWithContext(
		ctx,
		"POST",
		"https://github.com/login/device/code",
		strings.NewReader(data.Encode()),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to request device code: %w", err)
	}
	defer func() {
		_ = resp.Body.Close()
	}()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("GitHub returned status %d: %s", resp.StatusCode, string(body))
	}

	var deviceCode DeviceCodeResponse
	if err := json.Unmarshal(body, &deviceCode); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return &deviceCode, nil
}

// PollForAccessToken polls GitHub for the access token after user authorizes
// Returns the access token, or an error indicating the current state
func (c *OAuthClient) PollForAccessToken(
	ctx context.Context,
	deviceCode string,
) (*AccessTokenResponse, error) {
	data := url.Values{}
	data.Set("client_id", c.clientID)
	data.Set("device_code", deviceCode)
	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")

	req, err := http.NewRequestWithContext(
		ctx,
		"POST",
		"https://github.com/login/oauth/access_token",
		strings.NewReader(data.Encode()),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to poll for token: %w", err)
	}
	defer func() {
		_ = resp.Body.Close()
	}()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var tokenResp AccessTokenResponse
	if err := json.Unmarshal(body, &tokenResp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	// Check for error responses
	switch tokenResp.Error {
	case "authorization_pending":
		return nil, ErrAuthorizationPending
	case "slow_down":
		return nil, ErrSlowDown
	case "expired_token":
		return nil, ErrExpiredToken
	case "access_denied":
		return nil, ErrAccessDenied
	case "":
		// Success - we have a token
		if tokenResp.AccessToken == "" {
			return nil, errors.New("received empty access token")
		}
		return &tokenResp, nil
	default:
		return nil, fmt.Errorf("OAuth error: %s", tokenResp.Error)
	}
}

File: frontend/src/lib/api/oauth.ts
// Copyright (C) 2025 Austin Beattie
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import { fetchWithAuth } from "./fetch";

export interface DeviceFlowResponse {
	userCode: string;
	verificationUri: string;
	expiresIn: number;
	interval: number;
	deviceCode: string;
}

export interface PollResponse {
	status: "pending" | "complete" | "expired" | "denied";
	githubUsername?: string;
}

/**
 * Start the GitHub OAuth Device Flow.
 * Returns device code info including the verification URL to open.
 */
export async function startDeviceFlow(fetchImpl?: typeof fetch): Promise<DeviceFlowResponse> {
	const response = await fetchWithAuth(
		"/api/oauth/device",
		{
			method: "POST",
		},
		fetchImpl
	);

	if (!response.ok) {
		const error = await response
			.json()
			.catch(() => ({ error: "Failed to start GitHub authorization" }));
		throw new Error(error.error || "Failed to start GitHub authorization");
	}

	return response.json();
}

/**
 * Poll for the OAuth access token.
 * Call this repeatedly until status is "complete", "expired", or "denied".
 */
export async function pollForToken(
	deviceCode: string,
	fetchImpl?: typeof fetch
): Promise<PollResponse> {
	const response = await fetchWithAuth(
		"/api/oauth/poll",
		{
			method: "POST",
			headers: {
				"Content-Type": "application/json",
			},
			body: JSON.stringify({ deviceCode }),
		},
		fetchImpl
	);

	if (!response.ok) {
		const error = await response
			.json()
			.catch(() => ({ error: "Failed to check authorization status" }));
		throw new Error(error.error || "Failed to check authorization status");
	}

	return response.json();
}

/**
 * Run the complete OAuth Device Flow.
 * Opens the verification URL in a new window and polls until complete.
 *
 * @param onStatusChange - Called with status updates during the flow
 * @returns The GitHub username on success
 * @throws Error if the flow fails, expires, or is denied
 */
export async function runOAuthFlow(
	onStatusChange?: (status: "starting" | "waiting" | "complete" | "error", message?: string) => void
): Promise<string> {
	onStatusChange?.("starting", "Starting GitHub authorization...");

	// Start the device flow
	const deviceFlow = await startDeviceFlow();

	// Open the verification URL with the code pre-filled
	const verificationUrl = `${deviceFlow.verificationUri}?user_code=${deviceFlow.userCode}`;
	window.open(verificationUrl, "_blank", "noopener,noreferrer");

	onStatusChange?.("waiting", `Enter code: ${deviceFlow.userCode}`);

	// Poll for completion
	const pollInterval = Math.max(deviceFlow.interval, 5) * 1000; // At least 5 seconds
	const expiresAt = Date.now() + deviceFlow.expiresIn * 1000;

	while (Date.now() < expiresAt) {
		await new Promise((resolve) => setTimeout(resolve, pollInterval));

		const pollResult = await pollForToken(deviceFlow.deviceCode);

		switch (pollResult.status) {
			case "complete":
				onStatusChange?.("complete", `Connected as ${pollResult.githubUsername}`);
				return pollResult.githubUsername!;

			case "expired":
				throw new Error("Authorization expired. Please try again.");

			case "denied":
				throw new Error("Authorization was denied.");

			case "pending":
				// Continue polling
				break;
		}
	}

	throw new Error("Authorization timed out. Please try again.");
}
